packages/pipeline/src/feature-branches.ts

```diff
@@ -0,0 +1,361 @@
+import { LambdaFunction } from "aws-cdk-lib/aws-events-targets";
+import { Effect, PolicyStatement } from "aws-cdk-lib/aws-iam";
+import { Code, Function, Runtime } from "aws-cdk-lib/aws-lambda";
+import { CodePipelineProps } from "aws-cdk-lib/pipelines";
+import { NagSuppressions } from "cdk-nag";
+import { Construct } from "constructs";
+
+export interface FeatureBranchesProps
+  extends Pick<
+    CodePipelineProps,
+    "codeBuildDefaults" | "dockerEnabledForSynth"
+  > {
+  /**
+   * Branch name prefixes
+   * Any branches created matching this list of prefixes will create a new pipeline and stack.
+   *
+   * @example
+   * // Creates a new pipeline and stack for any branch
+   * new PDKPipeline(this, 'PDKPipeline', {
+   *   repositoryName: 'my-repo',
+   *   branchNamePrefixes: [''],
+   * }
+   * @example
+   * // Creates a new pipeline and stack for any branch starting with 'feature/' or 'fix/'
+   * new PDKPipeline(this, 'PDKPipeline', {
+   *   repositoryName: 'my-repo',
+   *   branchNamePrefixes: ['feature/', 'fix/'],
+   * }
+   * @example
+   * // Disables feature branches (default)
+   * new PDKPipeline(this, 'PDKPipeline', {
+   *   repositoryName: 'my-repo',
+   *   branchNamePrefixes: [], // or simply exclude this line
+   * }
+   * @default undefined
+   */
+  readonly branchNamePrefixes: string[];
+
+  /**
+   * The directory to run cdk synth from.
+   */
+  readonly cdkSrcDir: string;
+
+  /**
+   * The CodeCommit repository.
+   */
+  readonly codeRepository: IRepository;
+
+  /**
+   * Default branch.
+   */
+  readonly defaultBranchName: string;
+}
+
+export class FeatureBranches extends Construct {
+  constructor(scope: Construct, id: string, props: FeatureBranchesProps) {
+    super(scope, id);
+
+    const createFeatureBranchProject = new Project(
+      this,
+      "CreateFeatureBranchProject",
+      {
+        description: "Build project to deploy feature branch pipelines",
+        source: Source.codeCommit({ repository: props.codeRepository }),
+        environment: {
+          buildImage: LinuxBuildImage.STANDARD_6_0,
+          computeType: props.codeBuildDefaults?.buildEnvironment?.computeType
+            ? props.codeBuildDefaults.buildEnvironment.computeType
+            : ComputeType.SMALL,
+          privileged: props.dockerEnabledForSynth,
+        },
+        buildSpec: BuildSpec.fromObjectToYaml({
+          version: "0.2",
+          phases: {
+            install: {
+              commands: [
+                "npm install -g aws-cdk",
+                "yarn install --frozen-lockfile || npx projen && yarn install --frozen-lockfile",
+              ],
+            },
+            build: {
+              commands: [
+                "npx nx run-many --target=build --all",
+                `cd ${props.cdkSrcDir}`,
+                "npx cdk synth",
+                "npx cdk deploy --require-approval=never",
+              ],
+            },
+          },
```

Comment:
The user can pass custom install/build commands to `PDKPipeline`, and so I think we want to honour those here too?

Reply:
I think it will work to replace only `"npx nx run-many --target=build --all"` with `commands` from `PDKPipeline`. The user wouldn't pass installation commands to `commands` in `synthShellStepPartialProps`.

Reply:
You can pass `installCommands` to `synthShellStepPartialProps`, which is particularly useful if you're using a different package manager to yarn, like pnpm :) So I think we definitely want to honour what's configured in `PDKPipeline` where possible.

Reply:
You can pass `installCommands` to `synthShellStepPartialProps`, which is particularly useful if you're using a different package manager to yarn, like pnpm :) So I think we definitely want to honour what's configured in `PDKPipeline` where possible.